src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBioMartRetriever.java

```diff
@@ -0,0 +1,247 @@
+        }
+    }
+
+    // Updates BioMart XML Query with Microarray type information.
+    private String getMicroarrayTypesQuery(String bioMartSpeciesName) {
+        return this.getDataURL() + microarrayTypesBaseQuery.replace(BIOMART_SPECIES_NAME_PLACEHOLDER, bioMartSpeciesName);
+    }
+
+    /**
+     * Query BioMart for Microarray 'identifiers' associated with the microarray type and write the data to a file.
+     * @param biomartSpeciesName - String of species name in BioMart format (eg: hsapiens)
+     * @param biomartQueryFilePath - String, Path to the BioMart XML Query file
+     * @param biomartDataType - String, Type of data (either UniProt or Microarray) that will be queried for from BioMart
+     * @param fileSuffix - String that will be used to make filename (either _uniprot or _microarray) that will hold associated data.
+     * @throws IOException - Thrown when unable to write data to file.
+     */
+    private void queryBioMartAndStoreData(String biomartSpeciesName, String biomartQueryFilePath, String biomartDataType, String fileSuffix) throws IOException {
+
+        Set<String> biomartResponseLines = new HashSet<>();
+        try {
+            logger.info("Retrieving data associated with microarray type " + biomartDataType);
+            biomartResponseLines = queryBioMart(getBioMartIdentifierQuery(biomartQueryFilePath, biomartSpeciesName, biomartDataType), 0);
+        } catch (Exception e) {
+            logger.error("Unable to retrieve data associated with microarray type " + biomartDataType, e);
+            e.printStackTrace();
+        }
+
+        String biomartFilename = this.destination + biomartSpeciesName + fileSuffix;
+        storeBioMartData(biomartFilename, biomartResponseLines);
+
+    }
+
+    /**
+     * Write response from BioMart to file in the format 'BioMartSpeciesName_microarray' or 'BioMartSpeciesName_uniprot'.
+     * @param biomartFilename - String, Data from BioMart is saved to this filename.
+     * @param biomartResponseLines - Set<String>, All lines of response from BioMart.
+     * @throws IOException - Thrown when unable to write data to file.
+     */
+    private void storeBioMartData(String biomartFilename, Set<String> biomartResponseLines) throws IOException {
+
+        Set<String> biomartIdentifierLines = new HashSet<>();
+        for (String biomartResponseLine : biomartResponseLines) {
+            List<String> tabSplit = Arrays.asList(biomartResponseLine.split("\t"));
+            if (lineContainsIdentifier(tabSplit)) {
+                biomartIdentifierLines.add(biomartResponseLine + "\n");
+            }
+        }
+
+        // Write data to file.
+        for (String biomartIdentifierLine : biomartIdentifierLines) {
+            Files.write(
+                    Paths.get(biomartFilename),
+                    biomartIdentifierLine.getBytes(),
+                    StandardOpenOption.APPEND, StandardOpenOption.CREATE
+            );
+
+        }
```

Comment:
Why not write each line directly to the file, instead of adding them to `biomartIdentifierLines` first? Is there a reason for having two loops?

Reply:
Hmm I think that is actually an artifact from when the BioMart query would have the final line of the response as `[success]`. I'd first capture all values and verify the string existed in the response. Now that 'success' is not needed (say for cases where swissprot had a response but trembl didn't), I should probably refactor this to just write each line. Nice catch!

Reply:
On second reading, I notice that you're populating a Set. I guess having two loops could be useful to help eliminate duplicates (since you can't have duplicate values in a Set). Do you know if duplicates are possible, or an issue (if they do occur)?

Reply:
Duplicates shouldn't be possible when the response line has the 4th column (the microarray/uniprot identifier), but often the lines don't have that 4th column, which I suspect might have duplicates. The files are a bit smaller than what is produced via Perl (which I think is due to these theoretical duplicates). So ya, for now I'll keep it this way. 

The only real downside with waiting for the entire response to be received is that it makes it harder to interrogate if BioMart is hanging or not. Some of the responses take 20-30 minutes to be downloaded (looking at you Human microarray `huex`), and since it isn't being written on a per-line basis, it can be difficult to tell if it is downloading or hanging.

Reply:
If you process each line as it comes back, you could log progress evern _n_ lines (where n=1000, or something like that). Would that help determine if it's slow or hung/stuck?

Reply:
Yep, I'll do that

Reply:
Ok but _only_ if it would be useful. I don't want this PR to drag on forever and there's really nothing _wrong_ with the code as it's currently written.

Reply:
So as a result of the numerous refactorings, the 'queryBioMart' method gets the data but doesn't write to the file until the data is returned and then written to file. I think its a little too large of a refactoring for minimal gains, so for now I will leave that aspect as is. I added in logging though that logs status every 10000 lines